docs/design/2024-10-21-convert-partition.md

```diff
@@ -0,0 +1,198 @@
+  - Pro: easier to create a table that would be allowed to use for CONVERT TABLE TO PARTITION, including getting index ids matching.
+  - Con: extra work/time/resources during DDL operation.
+- Should we require "local" indexes on the non-partitioned table matching the partitioned tables global indexes? At least no need to match the internal index ids.
+  - Pro: at least there are no duplicate entries within the non-partitioned table.
+  - Con: it will just be dropped, and each row will be inserted and checked into the global indexes anyway, so not technically needed.
+
+Discussion:
+- @mjonss propose to require all indexes on the non-partitioned table t2 to match all non-global indexes of the partitioned table t.
+- @mjonss propose to not require t2 having indexes that matches global indexes of t.
+
+Note: Oracle supports `[{INCLUDING|EXCLUDING} INDEXES]` as option for also exchanging the local indexes, which we could use for including or excluding the Global Index to be converted to local indexes.
+
+#### 3) Move data from a non-partitioned table and replace an existing partition
+`ALTER TABLE t CONVERT TABLE t2 TO PARTITION p` basically the same syntax as 2), but only needing the partition name, and it will replace the data in that partition, so it also will work for KEY/HASH partitioned tables.
+MariaDB does not support this.
+It might be risky to support this, since it would efficiently TRUNCATE the existing partitions data and replace it with the data from t2.
+
+Discussion:
+- @mjonss propose to not support it, but rely on the following case instead + drop the "exchanged" table if the data is truly not needed.
+
+#### 4) Swap the data between a non-partitioned table with a partition.
+Logically the same as `ALTER TABLE t EXCHANGE PARTITION p WITH TABLE t2 [{WITH|WITHOUT} VALIDATION]`
+
+But currently the `EXCHANGE PARTITION` implementation in TiDB allows for concurrent read and write to both the exchanged table and partition, making it complex to support Global Index, where there will need to be a DDL state where one session, S1, sees the new t2 (original p) and the new p (original t2), while another session, S2, might see the original t2 and the original p. Meaning that S2 still needs to double write/update the Global Index of t, when writing to t2 and the S1 also needs to duble write/update the "local" index of t2 when writing to t/p, to keep both table's indexes consistent in the different views of S1 and S2. This makes the DDL extra complex and risky to implement and hard to test.
+Having a way to only allow access to the data of both t/p and t2 through table t during the DDL operation would make it less complex, fewer error handling conditions and also having less performance impact, due to less indexes to keep updated at the same time.
+
+So the question is if we should extend the syntax for `EXCHANGE PARTITION` for this new restriction and support Global Index or have a new syntax?
```

Comment:
Maybe let `exchange partition` same as MySQL and only supports the feature MySQL had.